controllers-films-users

Eugenpak opened the pull request:
controllers-films-users

Eugenpak committed:
Создать пакет model: Film, User.

Eugenpak committed:
Создать пакет controller: FilmController, UserController, RestExceptionHandler.

Eugenpak committed:
Создать пакет exception: ErrorMessage, NotFoundException, ValidationException.

Eugenpak committed:
Создать тесты: FilmControllerTest, UserControllerTest, FilmorateApplicationTests.

Eugenpak committed:
Изменить FilmorateApplication.

Eugenpak committed:
Checkstyle v-1

Eugenpak committed:
Checkstyle v-2

Eugenpak committed:
Checkstyle v-3

Eugenpak committed:
Checkstyle v-4

Eugenpak committed:
Checks tests v-1

Eugenpak committed:
Checkstyle v-5

Eugenpak committed:
Checkstyle v-6

Eugenpak committed:
Checks tests v-2

Eugenpak committed:
Checks tests v-3

Eugenpak committed:
Checks tests v-4

Eugenpak committed:
Checks tests v-5

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,29 @@
+import java.util.Date;
 
 /**
  * Film.
```
Такого рода комментарии лучше удалять. Смысла в них, в общем-то, нет. 🙂

Eugenpak replied:
Согласен. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,29 @@
-@Setter
+@Data
+@Builder
+@Slf4j
```
Логирование здесь не используется, поэтому библиотеку можно отключить.

Eugenpak replied:
Согласен. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,26 @@
+
+@Data
+@Builder
+@Slf4j
```
и здесь

Eugenpak replied:
Согласен. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,113 @@
+        }
+        if (newFilm.getName().isBlank()) {
+            log.error("ValidationException name");
+            throw new ValidationException("Описание должно содержать до 200 символов");
```
Описание должно содержать до 200 символов" -->   ("Название не может быть пустым"

Eugenpak replied:
Согласен. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,113 @@
+            log.warn("ValidationException releaseDate");
+            throw new ValidationException("Дата релиза — не раньше 28 декабря 1895 года.");
+        }
+        if (newFilm.getDuration() < 0) {
```
Вообще говоря, очень похожий набор проверок что здесь, что при добавлении фильма. 
Так и напрашивается провести небольшой рефакторинг и выделить в отдельный метод. 

На Ваше усмотрение.

Eugenpak replied:
Согласен. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,112 @@
+        return ++currentMaxId;
+    }
+
+    private User getUserEmail(String email) {
```
Как я понял, была идея проверять электронную почту на уникальность. Правда, она так и не была реализована, метод нигде не применяется. 🙂

На Ваше усмотрение. 
Для проверки уникальности электронного адреса в класс можно было бы добавить поле 

`private final Set emailUniqSet = new HashSet<>();`
Оно содержит набор  уникальных адресов, это будет такая же логика как ИНДЕКСы в БД .
Легко искать и проверять уникальность.
.-------------------------------------------------------

Соответсвенно сохранение будет выглядеть как-то так:
```java
public User addUser(User user) {
		final String email = user.getEmail();
		if (emailUniqSet.contains(email)) {
			throw new DataAlreadyExistException("Email: " + email + " already exists");
		}
		user.setId(++counter);
		users.put(user.getId(), user);
		emailUniqSet.add(email);
		return user;
	}
```

Eugenpak replied:
Уважаемый Александр! Проверка электронной почты было реализовано мной. Вот её реализация: 
//-----------------------------------------------------------------------------------
if (getUserEmail(newUser.getEmail()) != null && 
                getUserEmail(newUser.getEmail()).getId().longValue() != newUser.getId().longValue()) {
            log.error("ValidationException email");
            throw new ValidationException("Этот имейл уже используется");
}
//-----------------------------------------------------------------------------------
При проверки автоматических тестов и проверка в Postman (сценарий "PUT User update unknow") не проходил тест. Был вынужден отказаться от этой проверки, чтобы сдать ФЗ на проверку. Ваше решение тоже хорошее.

avfyodorov replied:
отлично,  вопрос снят ))

avfyodorov reviewed (commented):
Добрый день, Евгений! 

Хорошая работа, от меня всего пара уточнений.

Eugenpak committed:
Исправить: https://github.com/Eugenpak/java-filmorate/pull/1#discussion_r1957280718
           https://github.com/Eugenpak/java-filmorate/pull/1#discussion_r1957280882
           https://github.com/Eugenpak/java-filmorate/pull/1#discussion_r1957281007
           https://github.com/Eugenpak/java-filmorate/pull/1#discussion_r1957281562
           https://github.com/Eugenpak/java-filmorate/pull/1#discussion_r1957281966

avfyodorov reviewed (commented):
Добрый вечер, Евгений!

Всё в порядке, замечаний нет. 
Работа принята.
